crates/runtime/src/instance/allocator/pooling/memory_pool.rs

```diff
@@ -393,10 +541,185 @@ impl Drop for MemoryPool {
     }
 }
 
+/// The index of a memory allocation within an `InstanceAllocator`.
+#[derive(Clone, Copy, Debug, Eq, PartialEq, PartialOrd, Ord)]
+pub struct StripedAllocationIndex(u32);
+
+impl StripedAllocationIndex {
+    fn from_unstriped_slot_index(index: MemoryAllocationIndex, num_stripes: usize) -> Self {
+        let num_stripes: u32 = num_stripes.try_into().unwrap();
+        Self(index.0 / num_stripes)
+    }
+
+    fn as_unstriped_slot_index(self, stripe: usize, num_stripes: usize) -> MemoryAllocationIndex {
+        let num_stripes: u32 = num_stripes.try_into().unwrap();
+        let stripe: u32 = stripe.try_into().unwrap();
+        MemoryAllocationIndex(self.0 * num_stripes + stripe)
+    }
+}
+
+#[derive(Clone, Debug)]
+struct SlabConstraints {
+    max_memory_bytes: usize,
+    num_slots: usize,
+    expected_slot_bytes: usize,
+    num_pkeys_available: usize,
+    guard_bytes: usize,
```

Comment:
If some 32-bit user requests some kind memory configuration requiring 64-bit-sized addresses, how does that even work on a 32-bit machine? Don't we want to fail them early? Perhaps with a more descriptive error than an `unwrap()`?

Reply:
(Part of the reason this is readable is because of this "use `usize` everywhere" choice. When I switch to `u64` there are 40+ other places that will need to be fixed up with `try_into` or `as`).

Reply:
Definitely making a request on a 32-bit machine that can't be satisfied should return an error, but IMO it should be a normal error instead of a panic. Also to clarify I don't think everything should change to `u64`, only this structure. The `SlotLayout` structure should definitely use `usize` as it's tailored for the current machine, and my hope would be that `calculate` is the only location where `u64` is translated to `usize`, fallibly

Reply:
Ok, I misunderstood... for some reason I thought we were trying to refactor SlabLayout. Let me try this...